cc-wrapper: enable frame pointers by default.

eduarrrd opened the pull request:
Allow easy system-level profiling with e.g. Linux perf or bpf(trace).

Frame pointers are enabled by default in Fedora, Ubuntu, and Arch Linux.

---

Profiling applications in NixOS is quite expensive due to the size of closures of basic but widely-used libraries. Full-system profiling is even harder as it effectively requires re-compiling the world with ~0 caching.

By enabling it by default, we make this trivially possible (except for the packages that explicitly specify `-fomit-frame-pointers`, of course).

Due to these hurdles NixOS itself is hard to profile. I wouldn't be surprised if we find some low-hanging fruit once the infra is in place.

The change as such is not terribly controversial based on 
* https://discourse.nixos.org/t/nixos-frame-pointer-enable/39554 or
* https://github.com/NixOS/nixpkgs/pull/216341

... though there isn't a whole lot of discussion.

On the outside world part, from the big distributions, Fedora had it enabled the longest, and it survived re-evaluation. Ubuntu decided to add it to their LTS build so there's some commitment there. Arch still has it after >1y.

---

This PR is my best guess at how to actually accomplish this. Certainly looking for feedback here.

My rather simple test was to compare `find path/to/nixpkgs` built from this PR against the one on master, using `perf record -F 9999 -ag find ...`. It compiled and the perf reports looked as intended. Naturally the impact of this PR is much larger ... too large to test myself on my system.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

dramforever commented:
I am not familiar with how other distributions do this, but what about architectures with less number of registers, most notably `i*86`? The tradeoffs would be *very different* there.

I wonder what Ubuntu/Fedora do with `i*86` multilib. Maybe this shouldn't be enabled for every platform, but with an opt-in/opt-out list?

riking commented:
Shipping a default of "enable frame pointers if target_uname_m != i386" seems reasonable to me.

eduarrrd committed:
cc-wrapper: enable frame pointers by default.

Allow easy system-level profiling with e.g. Linux perf or bpf(trace).

Profiling applications in NixOS is quite expensive due to the size of
closures of basic but widely-used libraries. Full-system profiling is
even harder as it effectively requires re-compiling the world with ~0
caching.

By enabling it by default, we make this trivially possible (except for
the packages that explicitly specify -fomit-frame-pointers, of course)

Make an exception for 32bit x86 since it only has 8 general purpose
registers.

Frame pointers are enabled by default in Fedora, Ubuntu, and Arch Linux.

RossComputerGuy reviewed (commented):
This looks fine to me, I don't really have any objections but let's hold off after 25.05's branch off since I'd consider this a possible breaking change. I'm not sure how many packages might not compile if we merge this now.

RossComputerGuy commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 399014 --package stdenv --package util-linux --package hello`
Commit: `8eed8fb1a5041849aa84605604b106f2e79e6a32`

---
### `aarch64-linux`
<details>
  <summary>:x: 9 packages failed to build:</summary>
  <ul>
    <li>util-linux</li>
    <li>util-linux.bin (util-linux.bin.bin, util-linux.bin.debug, util-linux.bin.dev, util-linux.bin.lib, util-linux.bin.login, util-linux.bin.man, util-linux.bin.mount, util-linux.bin.swap)</li>
    <li>util-linux.debug (util-linux.debug.bin, util-linux.debug.debug, util-linux.debug.dev, util-linux.debug.lib, util-linux.debug.login, util-linux.debug.man, util-linux.debug.mount, util-linux.debug.swap)</li>
    <li>util-linux.dev (util-linux.dev.bin, util-linux.dev.debug, util-linux.dev.dev, util-linux.dev.lib, util-linux.dev.login, util-linux.dev.man, util-linux.dev.mount, util-linux.dev.swap)</li>
    <li>util-linux.lib (util-linux.lib.bin, util-linux.lib.debug, util-linux.lib.dev, util-linux.lib.lib, util-linux.lib.login, util-linux.lib.man, util-linux.lib.mount, util-linux.lib.swap)</li>
    <li>util-linux.login (util-linux.login.bin, util-linux.login.debug, util-linux.login.dev, util-linux.login.lib, util-linux.login.login, util-linux.login.man, util-linux.login.mount, util-linux.login.swap)</li>
    <li>util-linux.man (util-linux.man.bin, util-linux.man.debug, util-linux.man.dev, util-linux.man.lib, util-linux.man.login, util-linux.man.man, util-linux.man.mount, util-linux.man.swap)</li>
    <li>util-linux.mount (util-linux.mount.bin, util-linux.mount.debug, util-linux.mount.dev, util-linux.mount.lib, util-linux.mount.login, util-linux.mount.man, util-linux.mount.mount, util-linux.mount.swap)</li>
    <li>util-linux.swap (util-linux.swap.bin, util-linux.swap.debug, util-linux.swap.dev, util-linux.swap.lib, util-linux.swap.login, util-linux.swap.man, util-linux.swap.mount, util-linux.swap.swap)</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 2 packages built:</summary>
  <ul>
    <li>hello</li>
    <li>stdenv</li>
  </ul>
</details>

RossComputerGuy commented:
cryptsetup tests failed, didn't get a full look at it.

RossComputerGuy commented:
I think it was just a flaky test. I ran this on my 128 core Ampere Altra Max system so it's likely hard to reproduce this problem.

alyssais commented on pkgs/build-support/cc-wrapper/default.nix:
```diff
@@ -767,8 +767,12 @@ stdenvNoCC.mkDerivation {
       export hardening_unsupported_flags="${concatStringsSep " " ccHardeningUnsupportedFlags}"
     ''
 
+    + optionalString (!targetPlatform.isx86_32) ''
```
Could the reason for this check please be explained in a comment?

dramforever replied:
Maybe not just a comment. Maybe "should enable frame pointers by default" should go in `lib.platforms`?

alyssais replied:
Just checking because I don't know the architecture specifics: this reasoning applies only to x86, and not to other 32-bit architectures?

alyssais reviewed (changes requested):
Would be nice if the commit message included all the details you put in the PR body.

alyssais commented:
Semes fine to try.

misuzu commented:
This likely broke native riscv64-linux
```
% nix-build --no-out-link . -A pkgs.hello --eval-system riscv64-linux
<..................>
error: build of '/nix/store/qc1jfcxn5j2jwxw6k2h5iki7xc5b92di-expand-response-params.drv' on 'ssh://builder-vf2' failed: builder for '/nix/store/qc1jfcxn5j2jwxw6k2h5iki7xc5b92di-expand-response-params.drv' failed with exit code 1;
       last 9 log lines:
       > Running phase: unpackPhase
       > Running phase: patchPhase
       > Running phase: updateAutotoolsGnuConfigScriptsPhase
       > Running phase: configurePhase
       > no configure script, doing nothing
       > Running phase: buildPhase
       > gcc: error: unrecognized command-line option '-mno-omit-leaf-frame-pointer'; did you mean '-fno-omit-frame-pointer'?
       > /nix/store/159w6p1ziaa8hpf0p9x9y19dhvp8s873-bootstrap-stage1-stdenv-linux/setup: line 1807: pop_var_context: head of shell_variables not a function context
       > /nix/store/shkw4qm9qcw5sc5n1k5jznc83ny02r39-default-builder.sh: line 1: pop_var_context: head of shell_variables not a function context
       For full logs, run:
         nix log /nix/store/qc1jfcxn5j2jwxw6k2h5iki7xc5b92di-expand-response-params.drv
<..................>
```

LordGrimmauld commented:
Might want `targetPlatform.is32bit` instead
